CRM/Mailing/Service/ListUnsubscribe.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/**
+ * Apply a full range of `List-Unsubscribe` header options.
+ *
+ * @service civi.mailing.listUnsubscribe
+ * @link https://datatracker.ietf.org/doc/html/rfc8058
+ */
+class CRM_Mailing_Service_ListUnsubscribe extends \Civi\Core\Service\AutoService implements \Symfony\Component\EventDispatcher\EventSubscriberInterface {
+
+  public static function getMethods(): array {
+    return [
+      'mailto' => ts('Mailto'),
+      'http' => ts('HTTP(S) Web-Form'),
+      'oneclick' => ts('HTTP(S) One-Click'),
+    ];
+  }
+
+  public static function getSubscribedEvents() {
+    return [
+      '&hook_civicrm_alterMailParams' => ['alterMailParams', 1000],
+    ];
+  }
+
+  /**
+   * @see \CRM_Utils_Hook::alterMailParams()
+   */
+  public function alterMailParams(&$params, $context = NULL): void {
+    // FIXME: Flexmailer (BasicHeaders) and BAO (getVerpAndUrlsAndHeaders) separately define
+    // `List-Unsubscribe: <mailto:....>`. And they have separate invocations of alterMailParams.
+    //
+    // This code is a little ugly because it anticipates serving both code-paths.
+    // But the BAO path should be properly killed. Doing so will allow you cleanup this code more.
+
+    if (!in_array($context, ['civimail', 'flexmailer'])) {
+      return;
+    }
+
+    $methods = Civi::settings()->get('civimail_unsubscribe_methods');
+    if ($methods === ['mailto']) {
+      return;
+    }
+
+    if (!preg_match(';^<mailto:u\.(\d+)\.(\d+)\.(\w*)@(.*)>$;', $params['List-Unsubscribe'], $m)) {
```

Comment:
@totten testing at my end suggests two problems with this conditional check.

First, it's not checking for the possibility of a custom Localpart (Administer > CiviMail > Mail Accounts > Localpart).

Second, it's not checking for a custom/different VERP separator.

We have a Localpart value of `agweb` and a VERP separator of `-`. That results in ListUnsubscribe values like this: `<mailto:agweb+u-756713-2256...@...` (obviously trimmed).

This is quick/crude but perhaps this kind of approach is viable:
```
$regex_pattern = ";^<mailto:.*u\.(\d+)\.(\d+)\.(\w*)@(.*)>$;";
$verp_separator = Civi::settings()->get('verpSeparator');
if ($verp_separator !== ".") {
  $regex_pattern = str_replace("\.", "\\" . $verp_separator, $regex_pattern)
}
if (!preg_match($regex_pattern, $params['List-Unsubscribe'], $m)) {
  ...
```

Reply:
@artfulrobot oh that's nice. I didn't know `preg_quote` existed. Looking at some generated `mailto` strings in my testing, I think we'd need to also replace the `\.` in between the `(\d+)` elements with another `{$sep}`.

Reply:
Nice catch @johntwyman @artfulrobot. Updated both the 5.70 and 5.64 branches and included test-coverage. (The commits have the same implementation but slightly different tests.)